src/interfacers/EmonHubSDM120Interfacer.py

```diff
@@ -36,63 +36,73 @@ def __init__(self, name, device="/dev/modbus", baud=2400):
+
         self.next_interval = True
-        
-        # Only load module if it is installed        
-        try: 
+
+        # Only load module if it is installed
+        try:
             import sdm_modbus
-            self._log.info("Connecting to SDM120 device="+str(device)+" baud="+str(baud))
-            self._sdm = sdm_modbus.SDM120(device="/dev/modbus", baud=2400)
+            self._log.info("Connecting to SDM120 device=" + str(device) + " baud=" + str(baud))
+            self._sdm = sdm_modbus.SDM120(device=device, baud=int(baud))
             self._sdm_registers = sdm_modbus.registerType.INPUT
         except ModuleNotFoundError as err:
             self._log.error(err)
             self._sdm = False
-                    
+
 
     def read(self):
         """Read data and process
 
         Return data as a list: [NodeID, val1, val2]
 
         """
-        
-        if int(time.time())%self._settings['read_interval']==0:
-            if self.next_interval: 
+        # Read the following keys from the SDM120
+        read_keys = {
+            'voltage': ('V', 2),
+            'power_active': ('P', 2),
+            'power_factor': ('PF', 4),
+            'frequency': ('FR', 4),
+            'import_energy_active': ('E', 3),
+            'current': ('I', 3)
+        }
+
+        if int(time.time()) % self._settings['read_interval'] == 0:
+            if self.next_interval:
                 self.next_interval = False
 
                 c = Cargo.new_cargo()
                 c.names = []
                 c.realdata = []
                 c.nodeid = self._settings['nodename']
-             
+
                 if self._sdm and self._sdm.connected():
-                    r = self._sdm.read_all(self._sdm_registers)
-                    
+                    try:
+                        r = self._sdm.read_all(self._sdm_registers)
+                    except Exception as e:
+                        self._log.error("Could not read from SDM120: " + str(e))
                     # for i in r:
                     #     self._log.debug(i+" "+str(r[i]))
-                    
-                    self._log.debug("%.2f %.2f %.4f %.4f %.3f %.3f" % (r['voltage'],r['power_active'],r['power_factor'],r['frequency'],r['import_energy_active'],r['current']))
-                    
-                    c.names = ['sdm_V','sdm_P','sdm_PF','sdm_FR','sdm_E','sdm_I']
-                    c.realdata = [
-                      round(r['voltage'],2),
-                      round(r['power_active'],2),
-                      round(r['power_factor'],4),
-                      round(r['frequency'],4),
-                      round(r['import_energy_active'],3),
-                      round(r['current'],3)
-                    ]
-                    
-                    if len(c.realdata)>0:
+                    if r:
+                        try:
+                            for i in read_keys:
+                                if i in r:
```

Comment:
Instead of `for i in read_keys`, this should use something like `for i, (name, scale) in read_keys.items()`, rather than using `[0]` and `[1]` to index the tuple which is pretty unreadable.

Reply:
same as above, thanks!